### PR TITLE
feat(DFP): enable tls origination

### PR DESCRIPTION
**Pull request by gunnar-solo**
# Hey!  What's going on?
This is the follow-on to https://github.com/solo-io/gloo/pull/7817, in which the previously-added SslConfig is actually _consumed_.  Doing so, here, allows DynamicForwardProxy to use [TLS origination](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-field-config-cluster-v3-cluster-transport-socket).  This change is relatively simple, borrowing directly from [Envoy's Dynamic Forward Proxy](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/dynamic_forward_proxy_filter#config-http-filters-dynamic-forward-proxy) docs.

# How can I manually test/verify this?
```bash
# clean out kind cluster
kind delete cluster && kind create cluster

# build a local version of gloo edge
VERSION=1.0.0-ci make build-test-chart
VERSION=1.0.0-ci TAGGED_VERSION=1.0.0-ci make push-kind-images
helm install -n gloo-system gloo _test/gloo-1.0.0-ci.tgz --create-namespace

# install said built version
helm install -n gloo-system gloo _test/gloo-1.0.0-ci.tgz --create-namespace

# create a self-signed tls cert+key
# https://docs.solo.io/gloo-edge/latest/guides/security/tls/server_tls/
openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
   -keyout tls.key -out tls.crt -subj "/CN=petstore.example.com"
kubectl create secret tls gateway-tls --key tls.key \
   --cert tls.crt --namespace gloo-system
rm tls.crt tls.key

# apply gateway config for dfp
kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
kind: Gateway
metadata:
  labels:
    app: gloo
  name: gateway-proxy
  namespace: gloo-system
spec:
  bindAddress: '::'
  bindPort: 8080
  httpGateway:
    options:
      dynamicForwardProxy:
        sslConfig:
          secretRef:
            name: gateway-tls
            namespace: gloo-system
  proxyNames:
    - gateway-proxy
  ssl: false
  useProxyProto: false
EOF

# apply VS config for DFP
kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: test-static
  namespace: gloo-system
spec:
  virtualHost:
    domains:
    - '*'
    routes:
    - matchers:
      - prefix: /service
      options:
        regexRewrite:
          pattern:
            regex: ^/service/.*
          substitution: /get
        stagedTransformations:
          early:
            requestTransforms:
            - requestTransformation:
                transformationTemplate:
                  extractors:
                    destination:
                      header: :path
                      regex: /service/(.*)
                      subgroup: 1
                  headers:
                    x-destination:
                      text: '{{ destination }}'
      routeAction:
        dynamicForwardProxy:
          autoHostRewriteHeader: x-destination
EOF

# kubectl port-forward -n gloo-system deployments/gateway-proxy 8080:8080
# curl localhost:8080/service/52.200.117.68
# curl localhost:8080/service/httpbin.org
```

# "Regular Checklist"
- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go [hunk 1]
```diff
@@ -106,7 +107,38 @@ func generateCustomDynamicForwardProxyCluster(listenerCfg *dynamic_forward_proxy
 				TypedConfig: typedConfig,
 			},
 		},
-	}, nil
+	}
+
+	if sslConfig := listenerCfg.GetSslConfig(); sslConfig != nil {
```
Comment (gunnar-solo) [inline projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go:112]: using this cfg with some self-signed certs produces this error:

```
curl localhost:8080/service/52.1.93.201
upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: TLS error: 268435581:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED%
```

I'm not _totally_ in-the-loop on the validity of this, but it seems reasonable to me

File: projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go [hunk 2]
```diff
@@ -106,7 +107,38 @@ func generateCustomDynamicForwardProxyCluster(listenerCfg *dynamic_forward_proxy
...
+		// cfg := &envoyauth.UpstreamTlsContext{
```
Comment (gunnar-solo) [inline projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go:113]: as opposed to _this_ config, which uses a hardcoded `ca-certificates.crt`, which works perfectly fine

File: projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go [hunk 3]
```diff
@@ -106,7 +107,24 @@ func generateCustomDynamicForwardProxyCluster(listenerCfg *dynamic_forward_proxy
 				TypedConfig: typedConfig,
 			},
 		},
-	}, nil
+	}
+
+	if sslConfig := listenerCfg.GetSslConfig(); sslConfig != nil && params.Snapshot != nil {
```
Comment (gunnar-solo) [inline projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go:112]: for ease of review:  this is almost the entirety of the changes required for this feature.  The rest is codegen + a supporting test

**Comment by gunnar-solo** [comment]
/kick [generating unique route names should generate unique names for multiple proxies](https://console.cloud.google.com/cloud-build/builds;region=global/319681d7-8478-44dc-9ff0-82e7bce0771e?project=solo-public)

**Comment by gunnar-solo** [comment]
/kick [neither TAGGED_VERSION nor TEST_ASSET_ID are set](https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=36c4601e-a26d-4e12-a58a-4ebb518688b5)

**Comment by gunnar-solo** [comment]
/kick [exploded logs](https://console.cloud.google.com/cloud-build/builds;region=global/c791e26e-b20b-4c04-adb6-62d2f32304d5;step=10?project=solo-public)

**Review by elcasteel — DISMISSED** [review]

File: changelog/v1.14.0-beta12/7693-feature-complete.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+changelog:
+  - type: NEW_FEATURE
+    issueLink: https://github.com/solo-io/gloo/issues/7693
+    resolvesIssue: false
+    description: >-
+      Allow configured DynamicForwardProxies to use TLS Origination (https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-field-config-cluster-v3-cluster-transport-socket) via a newly exposed field
```
Comment (jenshu) [inline changelog/v1.14.0-beta12/7693-feature-complete.yaml:6]: nit: can you include the field name here?

File: projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go (hunk 3 above)
Comment (jenshu) [inline projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go:112]: would params.Snapshot ever be nil? i think it's fine to check that here, but our other plugins seem to just use it without checking

Comment (gunnar-solo) [reply projects/gloo/pkg/plugins/dynamic_forward_proxy/plugin.go:112]: great question!

...probably not.  However, I did manage to _make_ it be nil when I was writing my test

**Review by elcasteel — APPROVED** [review]

**Comment by gunnar-solo** [comment]
/kick https://console.cloud.google.com/cloud-build/builds;region=global/a9b0a905-e9f8-401d-bf8f-4f83ad3f5946?project=solo-public

**Review by jenshu — APPROVED** [review]

**Comment by sam-heilbron** [comment]
/kick

https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=97f7d4a5-f94b-4952-a160-f7935f593493
```
- "test": [91m[1mSummarizing 1 Failure:[0m
Step #10 - "test": 
Step #10 - "test": [91m[1m[Fail] [0m[90mStaged Transformation [0m[0mno auth [0m[91m[1m[It] Can extract a substring from the body [0m
Step #10 - "test": [37m/workspace/gloo/test/e2e/staged_transformation_test.go:291[0m
Step #10 - "test": 
Step #10 - "test": [1m[91mRan 11 of 189 Specs in 58.747 seconds[0m
Step #10 - "test": [1m[91mFAIL![0m -- [32m[1m10 Passed[0m | [91m[1m1 Failed[0m | [33m[1m0 Pending[0m | [36m[1m178 Skipped[0m
Step #10 - "test": --- FAIL: TestE2e (62.78s)
Step #10 - "test": FAIL
```

This is the first time I've seen this failure, so kicking to see if it's a flake

**Review by elcasteel — APPROVED** [review]

**Pull request merged**